### ISSUE TITLE
GRUB2 ordering is unreliable

### ISSUE THREAD

**Issue by ikeydoherty** [clearlinux/clr-boot-manager#62]
On certain systems we're seeing GRUB2 menus default to selecting index 0.
On a system with an existing kernel and an upgrade, an unattended reboot leads to booting the old
kernel.

In terms of usability, for some users they may never end up on the new kernel without explicitly selecting it.
Thus, we should alter GRUB2 to emit a single `/etc/grub.d` file, nuke old files, and use submenus for the "alternative boot entries".

This ensures that our default kernel, idx 0, is always selected first, and we stuff older/other entries under a different submenu structure.

Codewise, we'll make the GRUB2 implementation similar to the syslinux implementation.

**Comment by ikeydoherty** [comment]
Using `grub-submenus` I now get:
```
#
# DO NOT EDIT THIS FILE
#
# It is automatically generated by grub-mkconfig using templates
# from /etc/grub.d and settings from /etc/default/grub
#

### BEGIN /etc/grub.d/00_header ###
if [ -s $prefix/grubenv ]; then
  load_env
fi
if [ "${next_entry}" ] ; then
   set default="${next_entry}"
   set next_entry=
   save_env next_entry
   set boot_once=true
else
   set default="0"
fi

if [ x"${feature_menuentry_id}" = xy ]; then
  menuentry_id_option="--id"
else
  menuentry_id_option=""
fi

export menuentry_id_option

if [ "${prev_saved_entry}" ]; then
  set saved_entry="${prev_saved_entry}"
  save_env saved_entry
  set prev_saved_entry=
  save_env prev_saved_entry
  set boot_once=true
fi

function savedefault {
  if [ -z "${boot_once}" ]; then
    saved_entry="${chosen}"
    save_env saved_entry
  fi
}

function load_video {
  if [ x$feature_all_video_module = xy ]; then
    insmod all_video
  else
    insmod efi_gop
    insmod efi_uga
    insmod ieee1275_fb
    insmod vbe
    insmod vga
    insmod video_bochs
    insmod video_cirrus
  fi
}

if loadfont unicode ; then
  set gfxmode=auto
  load_video
  insmod gfxterm
  set locale_dir=$prefix/locale
  set lang=en_IE
  insmod gettext
fi
terminal_output gfxterm
insmod part_msdos 
insmod ext2
set root='hd0,msdos1'
if [ x$feature_platform_search_hint = xy ]; then
  search --no-floppy --fs-uuid --set=root --hint-bios=hd0,msdos1 --hint-efi=hd0,msdos1 --hint-baremetal=ahci0,msdos1  96b9a863-9ce3-4e18-87cb-6f50a9308025
else
  search --no-floppy --fs-uuid --set=root 96b9a863-9ce3-4e18-87cb-6f50a9308025
fi
insmod tga
background_image -m stretch /grub/splash.tga
if [ x$feature_timeout_style = xy ] ; then
  set timeout_style=menu
  set timeout=5
# Fallback normal timeout code in case the timeout_style feature is
# unavailable.
else
  set timeout=5
fi
### END /etc/grub.d/00_header ###

### BEGIN /etc/grub.d/10_com.solus-project ###
menuentry 'Solus 2017.01.01.0 (4.9.18-11.lts)' --class solus --class gnu-linux --class gnu --class os $menuentry_id_option 'solus-4.9.18-11.lts' {
	load_video
	insmod gzio
	insmod part_msdos 
insmod ext2
set root='hd0,msdos1'
if [ x$feature_platform_search_hint = xy ]; then
  search --no-floppy --fs-uuid --set=root --hint-bios=hd0,msdos1 --hint-efi=hd0,msdos1 --hint-baremetal=ahci0,msdos1  96b9a863-9ce3-4e18-87cb-6f50a9308025
else
  search --no-floppy --fs-uuid --set=root 96b9a863-9ce3-4e18-87cb-6f50a9308025
fi
	echo 'Loading Solus 2017.01.01.0 4.9.18 ...'
	linux /com.solus-project.lts.4.9.18-11 root=UUID=f3c9eeb8-17e5-42b2-972b-b47ce931f181 rd.luks.uuid=e725c7c3-49da-442f-94b0-ffc632481d9d quiet ro splash 
	echo 'Loading initial ramdisk'
	initrd /initrd-com.solus-project.lts.4.9.18-11
}
submenu 'Solus 2017.01.01.0 (alternative boot entries)' $menuentry_id_option 'com.solus-project-cbm-submenu' {
	menuentry 'Solus 2017.01.01.0 (4.9.17-9.lts)' --class solus --class gnu-linux --class gnu --class os $menuentry_id_option 'solus-4.9.17-9.lts' {
	load_video
		insmod gzio
	insmod part_msdos 
insmod ext2
set root='hd0,msdos1'
if [ x$feature_platform_search_hint = xy ]; then
  search --no-floppy --fs-uuid --set=root --hint-bios=hd0,msdos1 --hint-efi=hd0,msdos1 --hint-baremetal=ahci0,msdos1  96b9a863-9ce3-4e18-87cb-6f50a9308025
else
  search --no-floppy --fs-uuid --set=root 96b9a863-9ce3-4e18-87cb-6f50a9308025
fi
		echo 'Loading Solus 2017.01.01.0 4.9.17 ...'
		linux /com.solus-project.lts.4.9.17-9 root=UUID=f3c9eeb8-17e5-42b2-972b-b47ce931f181 rd.luks.uuid=e725c7c3-49da-442f-94b0-ffc632481d9d quiet ro splash 
		echo 'Loading initial ramdisk'
		initrd /initrd-com.solus-project.lts.4.9.17-9
	}
	menuentry 'Solus 2017.01.01.0 (4.9.17-9.lts)' --class solus --class gnu-linux --class gnu --class os $menuentry_id_option 'solus-4.9.17-9.lts' {
	load_video
		insmod gzio
	insmod part_msdos 
insmod ext2
set root='hd0,msdos1'
if [ x$feature_platform_search_hint = xy ]; then
  search --no-floppy --fs-uuid --set=root --hint-bios=hd0,msdos1 --hint-efi=hd0,msdos1 --hint-baremetal=ahci0,msdos1  96b9a863-9ce3-4e18-87cb-6f50a9308025
else
  search --no-floppy --fs-uuid --set=root 96b9a863-9ce3-4e18-87cb-6f50a9308025
fi
		echo 'Loading Solus 2017.01.01.0 4.9.17 ...'
		linux /com.solus-project.lts.4.9.17-9 root=UUID=f3c9eeb8-17e5-42b2-972b-b47ce931f181 rd.luks.uuid=e725c7c3-49da-442f-94b0-ffc632481d9d quiet ro splash 
		echo 'Loading initial ramdisk'
		initrd /initrd-com.solus-project.lts.4.9.17-9
	}
}
### END /etc/grub.d/10_com.solus-project ###

### BEGIN /etc/grub.d/10_linux ###

### END /etc/grub.d/10_linux ###

### BEGIN /etc/grub.d/20_linux_xen ###
### END /etc/grub.d/20_linux_xen ###

### BEGIN /etc/grub.d/30_os-prober ###
### END /etc/grub.d/30_os-prober ###

### BEGIN /etc/grub.d/40_custom ###
# This file provides an easy way to add custom menu entries.  Simply type the
# menu entries you want to add after this comment.  Be careful not to change
# the 'exec tail' line above.
### END /etc/grub.d/40_custom ###

### BEGIN /etc/grub.d/41_custom ###
if [ -f  ${config_directory}/custom.cfg ]; then
  source ${config_directory}/custom.cfg
elif [ -z "${config_directory}" -a -f  $prefix/custom.cfg ]; then
  source $prefix/custom.cfg;
fi
### END /etc/grub.d/41_custom ###
```

**Comment by ikeydoherty** [comment]
Now fixed in master, will appear in 1.5.2 release
